### ISSUE TITLE
How do you handle dynamic translations?

### ISSUE THREAD

**Issue by rssfrncs** [lingui/js-lingui#177]
Hi,

Is it possible to translate the example below?

```
{routes.map(route => <Trans>{route}</Trans>)}
```

If routes equals ['home', 'contact'] i'd expect two new keys to be extracted 'home' and 'contact'.

**Comment by tricoder42** [comment]
Yeah, first you need to use `i18nMark` to mark translatable strings and then pass it as an id either to `<Trans>` component or `i18n._` method:

```js
const routes = [i18nMark('home'), i18nMark('contact')]

routes.map(route => <Trans id={route} />)

```

**Comment by rssfrncs** [comment]
@tricoder42 great and thanks for the quick response. 

Can i suggest that this is added as a todo "to add to documentation"? I imagine that this is quite a common use-case and i couldn't find anything about it in the docs site.

**Comment by tricoder42** [comment]
It's documented in [reference docs](https://lingui.github.io/js-lingui/ref/lingui-react.html#i18nMark), but would be great to add it either to tutorial or how-to guide. It's a bit hidden in reference docs.

Maybe we just need to add a reference to a docs. Where were you looking when you were searching for this feature?

**Comment by rssfrncs** [comment]
@tricoder42 I'd say this could be made search friendly by simply including "dynamic translation key". I used the docs site top level search.

**Comment by tricoder42** [comment]
Good point! Thanks 👍

**Comment by rssfrncs** [comment]
@tricoder42 just to confirm my reasoning here, i18nMark purpose is for lingui babel to identify and extract it's called value? The implementation of i18nMark is `key => key` and does 'nothing'.

**Comment by tricoder42** [comment]
Exactly, `@lingui/babel-plugin-extract-messages` finds `i18nMark` function calls and extracts it and remove `i18nMark` function completely.

**Comment by HriBB** [comment]
But what if keys are not available at compile time? For example, I am loading a list of modules from the GraphQL server and render it in a drawer. How can I translate that?

**Comment by tricoder42** [comment]
@HriBB That's actually more complex and jsLingui doesn't support that at the moment.

If you know the messages, you could write a script which queries GraphQL server, extracts messages and then run `lingui compile`.
If you don't know the messages (e.g.: dynamic content on website), then you have to load message parser anyway. I need to figure out how to override default behaviour a force message parser to be included in production bundle as well.

**Comment by rssfrncs** [comment]
@HriBB @tricoder42 perhaps the server should be handling translation in this case?

**Comment by HriBB** [comment]
@tricoder42 it's similar to what I am doing now. I manually extract all the translations from GraphQL server, and build a `server/.cache/translations.js`, then include this folder in `srcPathDirs`. What if the `cli` had like a hook script, which would return a promise, and in this promise, user could add additional translations. You could include extra scripts, query the server, merge catalogs, ...

@rssfrncs perhaps :) Probably. I actually thought about that, but for now, I have only a handful of translations on the server, and would like to keep them in one place. I have a few dynamic modules, and all I need is to translate names of modules. But for bigger projects, I think the server should be handling translations.

**Comment by MartijnHols** [comment]
When using `i18n._` for dynamic ids, the extractor complains about missing message IDs. Should a different method be used?

**Comment by tricoder42** [comment]
@MartijnHols No, that should work actually. Could you please show an example?

**Comment by donatoaz** [comment]
Hi @HriBB , can you share a little bit more of your implementation? I am having a hard time visualizing what should be the content of your `server/.cache/translations.js`.

When you say manually, do you mean your dataset changes very infrequently? What if your changes were more frequent? is it something you'd be able to automate? Perhaps via a cronjob or anything periodical.

I have a very simple use case of a `categories` table that needs to be translatable (name and description columns).

@tricoder42 did the idea of hook scripts got incorporated somehow to the package? 

Thanks for the attention!
